### ISSUE TITLE
Typing for AST interface

### ISSUE THREAD

**Issue by cos** [kframework/kore#11]
Moving here the discussion about the type constraints on the AST interface as it's easier to track (and code is easier to read).

### 1. The current solution

```
trait AST[T <: AST[T]]

trait Node[T <: AST[T]] extends AST[T] {
  def children: Seq[T]
}

trait Pattern extends AST[Pattern]
```

Advantages:
 + good separation of concerns (separates AST-related functionality from Pattern-related functionality, allowing AST to be tested separately without "seeing" the Pattern interface)
 + good compatibility with Java (these Scala type parameters and bounds have direct equivalents in Java)

Disadvantage:
- More complex and verbose in usage (references to `Node` require type parameters: `Node[T]`)

###  2. Scala type members

```
trait AST {
  type T <: AST
}

trait Node extends AST {
  def children: Seq[T]
}

trait Pattern extends AST {
  type T = Pattern
}
```

Advantages:
 + good separation of concerns
 + `T` is not relevant/visible when using traits (references to `Node` **do not** require type parameters, they ca simply be `Node`)

Disadvantage:
- bad compatibility with Java (these Scala abstract type members **do not** have direct equivalents in Java, so, e.g., `children` will become a `Seq` of `Object`s)

If you think this is an alternative worth considering, I can make a diff showing more details.

###  3. Have `AST` "know" about `Pattern` 

```
trait AST

trait Node extends AST {
  def children: Seq[Pattern]
}

trait Pattern extends AST
```

Advantages:
 + there are no type parameters (references to `Node` are simply `Node`)
 + no compatibility issues with Java 

Disadvantage:
- bad separation of concerns. The way I see this, the `AST` is more abstract than the `Pattern`. If the `AST` interface has references to the `Pattern` interface, there is no way to write code that doesn't care/know that it is looking at `Pattern`s. I think this will make code somewhat uglier in the long run, but the simplicity of not having type parameters may be worth it.

**Comment by cos** [comment]
@grosu wrote:

> When you say "dependent" you mean in the type-dependent sense, or what I actually said in my previous message, which is:
> * get rid of all parameters T and _, and just have two basic traits, Ast and Pattern, with Pattern a subtype of Ast
> * replace Seq[_ <: T] by Seq[Pattern]
> * replace Node[T], etc., with just Node
> If this is what you meant, then is there anything we lose with this approach, except a generality which we currently do not need?

I mean what you said in the previous message. It is option 3 above.

**Comment by cos** [comment]
@grosu wrote:

> What separation of concerns and what compatibility with Java?  It is just complicated.  How does it give you more compatibility with Java than just dropping all parameters everywhere?

In terms of Java compatibility, options 1 and 3 are the same, while option 2 has a problem. And yes, option 3 is simpler.

**Comment by grosu** [comment]
@cos, thanks for doing this.

So, compatibility with Java is critical, because we are not going to rewrite the backend any time soon.  From that point of view, approach 2 falls.  Also, the generality of AST and separation of concerns related to Patterns are really not a big issue here.  Everybody knows that AST can be general and there are libraries for them, etc., but the whole point of our KORE implementation is that it is specific to our needs.  Also, simplicity of code is very important.  Several people from our group spend tens of hours to "fight" Scala's type system while developing the current code that Manasvi pushed, with T <: AST[T].  Some were even very furious.  That is telling something about the approach, too.  We cannot afford to have future generations of students stare at that code for days and not even know how to start.  In some sense, that's the reason why we switched from Haskell to Java a couple of years ago.

So I would personally go with 3 for now.  But before we decide, I would like to understand your concerns a bit better.  You said:

> bad separation of concerns. The way I see this, the AST is more abstract than the Pattern. If the AST
> interface has references to the Pattern interface, there is no way to write code that doesn't care/know 
> that it is looking at Patterns. I think this will make code somewhat uglier in the long run, but the 
> simplicity of not having type parameters may be worth it.

The fact that our ASTs are crafted for our Patterns is OK in my view.  It is the Patterns that really matter.  The AST view is only a convenience, so people can write more compact code.  There is still nothing to stop them from picking any generic AST library, or develop one in the style proposed by @ehildenb and @lucaspena , and then have Patterns refine that one.

But why do you think the code will be uglier?  People are only supposed to implement the Pattern interface, the AST interface will be there under the hood, "for free if you need it", won't it?  If they ever need to see Patterns as ASTs, then they can just use the `Node(c)` or `Leaf(l)` destructors whenever they please, or `n:Node`, etc., but they will not need to add code to the AST, will they?

**Comment by cos** [comment]
I'm reordering the questions to narrow down the crux of the argument.

> People are only supposed to implement the Pattern interface, the AST interface will be there under the hood, "for free if you need it", won't it? 

Yes, it will be for free if you need it.

> If they ever need to see Patterns as ASTs, then they can just use the Node(c) or Leaf(l) destructors whenever they please, or n:Node, etc., 

Yes, but just in principle -- the the last question.

> but they will not need to add code to the AST, will they?

No, they will not need to add code. 

> But why do you think the code will be uglier?

I mean uglier from the perspective of separation of concerns. For analogy, let's consider functions. Their main use (aside from recursive algorithms) is to separate code dealing with a specific issue from the rest of the code. Similarly, in our case, the AST interfaces could also be used to write code that doesn't know anything about ML. Yes, a careful user will only use the `Node` and `Leaf` destructors and methods and have very clean code. But there will be nothing in the type system that would stop him from also using methods of Pattern after accessing the children. Let's consider the following code:

```
def elements(n: Pattern) = { 
  // ... returns either children(n) or the flattened list of children if the node is assoc 
}

def orderedTraversal(n: AST): List[AST] = n match {
   case Node(c) => (c flatMap elements flatMap orderedTraversal) :+ n
   case _ => n
}
```

Aside from being an ugly recursive implementation, the code is also incorrect because `elements` may return the elements of an AC term in any order, breaking the contract of `orderedTraversal(n: AST)` that assumed we are only looking at the AST structure.

With option 1 or 2, this code would fail to compile because `c` would be of type `List[AST]`, not `List[Pattern]`. In order to make it compile, we would have to change the signature of the function to
`def orderedTraversal(n: Pattern): List[Pattern]`, which would make it clear that we're talking about `Pattern`s and need to be careful about any non-AST behaviors.

Bottom line, I don't have a strong opinion regarding this, just want to highlight that we are losing some separation/safety if we forgo the parametric types for AST. Also, note that in the 1 proposal for parametric types, only code making direct references to `AST/Node[...]` would be affected by its extra verbosity. If the user chooses to only "talk" about `Pattern`s, none of the extra verbosity would be visible.

**Comment by grosu** [comment]
@cos , anything still relevant here, or we can close?

**Comment by cos** [comment]
We made a decision so I think we can close.
